operator/e2e_testing/setup/shared_cluster.go

```diff
@@ -0,0 +1,483 @@
+					totalResources += len(resourceList.Items)
+				}
+			}
+
+			// Check pods
+			allPodsDeleted := true
+			nonSystemPods := 0
+			pods, err := scm.clientset.CoreV1().Pods("default").List(ctx, metav1.ListOptions{})
+			if err == nil {
+				for _, pod := range pods.Items {
+					if !isSystemPod(&pod) {
+						allPodsDeleted = false
+						nonSystemPods++
+					}
+				}
+			}
+
+			if allResourcesDeleted && allPodsDeleted {
+				return nil
+			}
+
+			if totalResources > 0 || nonSystemPods > 0 {
+				scm.logger.Debugf("⏳ Waiting for %d Grove resources and %d pods to be deleted...", totalResources, nonSystemPods)
+			}
+		}
+	}
+}
+
+// listRemainingResources lists remaining Grove resources for debugging
+func (scm *SharedClusterManager) listRemainingResources(ctx context.Context) {
+	resourceTypes := []struct {
+		group    string
+		version  string
+		resource string
+		name     string
+	}{
+		{"grove.io", "v1alpha1", "podcliquesets", "PodCliqueSets"},
+		{"grove.io", "v1alpha1", "podcliquescalinggroups", "PodCliqueScalingGroups"},
+		{"scheduler.grove.io", "v1alpha1", "podgangs", "PodGangs"},
+	}
+
+	for _, rt := range resourceTypes {
+		gvr := schema.GroupVersionResource{
+			Group:    rt.group,
+			Version:  rt.version,
+			Resource: rt.resource,
+		}
+
+		resourceList, err := scm.dynamicClient.Resource(gvr).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			scm.logger.Warnf("Failed to list %s: %v", rt.name, err)
+			continue
+		}
+
+		if len(resourceList.Items) > 0 {
+			resourceNames := make([]string, 0, len(resourceList.Items))
+			for _, item := range resourceList.Items {
+				resourceNames = append(resourceNames, fmt.Sprintf("%s/%s", item.GetNamespace(), item.GetName()))
+			}
+			scm.logger.Warnf("Remaining %s: %v", rt.name, resourceNames)
+		}
+	}
+}
+
+// GetClients returns the kubernetes clients for tests to use
+func (scm *SharedClusterManager) GetClients() (*kubernetes.Clientset, *rest.Config, dynamic.Interface) {
+	return scm.clientset, scm.restConfig, scm.dynamicClient
+}
+
+// GetRegistryPort returns the registry port for test image setup
+func (scm *SharedClusterManager) GetRegistryPort() string {
+	return scm.registryPort
+}
+
+// GetWorkerNodes returns the list of worker node names
+func (scm *SharedClusterManager) GetWorkerNodes() []string {
+	return scm.workerNodes
+}
+
+// IsSetup returns whether the shared cluster is setup
+func (scm *SharedClusterManager) IsSetup() bool {
+	return scm.isSetup
+}
+
+// Teardown cleans up the shared cluster
+func (scm *SharedClusterManager) Teardown() {
+	if scm.cleanup != nil {
+		scm.cleanup()
+		scm.isSetup = false
+	}
+}
+
+// setupRegistryTestImages sets up test images in the registry
+func setupRegistryTestImages(registryPort string, images []string) error {
+	if len(images) == 0 {
+		return nil
+	}
+
+	ctx := context.Background()
+
+	// Initialize Docker client
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
```

Comment:
Nit:

I'm very used to seeing `client` and thinking k8s client. Although the surrounding context and comment right above add this information, naming the import is also not a bad idea in my opinion.

```suggestion
	cli, err := dockerclient.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
```